runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -410,44 +435,119 @@ public void processWatermark2(Watermark mark) throws Exception {
   }
 
   @Override
-  public StreamTaskState snapshotOperatorState(
-      long checkpointId,
-      long timestamp) throws Exception {
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    // copy from AbstractStreamOperator
+    if (getKeyedStateBackend() != null) {
+      KeyedStateCheckpointOutputStream out;
+
+      try {
+        out = context.getRawKeyedOperatorStateOutput();
+      } catch (Exception exception) {
+        throw new Exception("Could not open raw keyed operator state stream for "
+            + getOperatorName() + '.', exception);
+      }
+
+      try {
+        KeyGroupsList allKeyGroups = out.getKeyGroupList();
+        for (int keyGroupIdx : allKeyGroups) {
+          out.startNewKeyGroup(keyGroupIdx);
 
-    StreamTaskState streamTaskState = super.snapshotOperatorState(checkpointId, timestamp);
+          DataOutputViewStreamWrapper dov = new DataOutputViewStreamWrapper(out);
 
-    if (sideInputStateBackend != null) {
-      // we have to manually checkpoint the side-input state backend and store
-      // the handle in the "user state" of the task state
-      HashMap<String, KvStateSnapshot<?, ?, ?, ?, ?>> sideInputSnapshot =
-          sideInputStateBackend.snapshotPartitionedState(checkpointId, timestamp);
+          // if (this instanceof KeyGroupCheckpointedOperator)
+          snapshotKeyGroupState(keyGroupIdx, dov);
 
-      if (sideInputSnapshot != null) {
-        @SuppressWarnings("unchecked,rawtypes")
-        StateHandle<Serializable> sideInputStateHandle =
-            (StateHandle) sideInputStateBackend.checkpointStateSerializable(
-                sideInputSnapshot, checkpointId, timestamp);
+          // We can't get all timerServices, so we just snapshot our timerService
```

Comment:
We don't have {{getInternalTimerServiceByName(String name)}} method. There maybe be a problem about invoking {{HeapInternalTimerService.startTimerService()}} by using null namespaceSerializer or null Trigger.